README.md
# C/C++ Comment Location Suggestions
This plugin uses the model structure from the paper "Where should I comment my code? A dataset and model for predicting locations that need comments" (https://homes.cs.washington.edu/~mernst/pubs/predict-comments-icse2020-abstract.html)



## PART 1: Creating deep learning model for the IntelliJ plugin:
If you want to train the model from scratch, you can do as follow:

### install requirements
```
pip install -r requirements.txt
```
### install training data and extract it 
```
wget https://groups.inf.ed.ac.uk/cup/comment-locator/data.tar.gz
tar -xf data.tar.gz
```
### training the model 
You can train the model with arguments as follow. You can also specify more arguments from below. 
There is also notebook version 'train_orig_ptl.ipynb' which you can run on Google Colab.
```
python train_orig.py --mode train  --data_train_path  path/to/data/split30_word_data/train/data.txt \
                 --data_validation_path path/to/data/split30_word_data/valid/data.txt \
                 --code_embeddings  path/to/data/code-big-vectors-negative300.bin \
```
command-line arguments 
```
--data_train_path, default="data/split30_word_data/train/data.txt",
--data_test_path, default="data/split30_word_data/test/data.txt",
--data_validation_path, default="data/split30_word_data/valid/data.txt",
--check_point_dir, default="model_save/", help="Directory for saving model during training",
--code_embeddings, default="data/code-big-vectors-negative300.bin", help="Pretrained embeddings",
--embed_size, default=300, type=int, help="Word token dimension, do not change it unless you change preembedding for words",
--max_blks, default=15, type=int, help="Max number of blocks in file",
--max_len_blk, default=15, type=int, help="Max size of a block",
--max_len_stmt, default=15, type=int, help="Max size of a line",
--vocab_size, default=5000, type=int, help="Whole vocabulary size",
--embedding_type, default="avg_wembed", help="Only average embeddings are implemented",

# Model related arguments #
--use_gpu, default=0, type=int, help="1 for GPU usage",
--lstm_hidden_size, default=800, type=int, help="")
--lstm_num_layers, default=1, help="")
--batch_size, default=64, type=int, help="Batch size for training"
--learning_rate, default=0.001, type=float)
--max_epochs, default=1, type=int)
--mode, default="train", help="train or predict")
--model_for_prediction, default="none", help="Path for model to predict"



```

### exporting Pytorch model to ONNX model
To be able to integrate the previous Pytorch model to Plugin, it needs to be converted ONNX format. Following command will will extract it as "model.onnx"

```
python onnx_export.py
```


## PART 2: Comment Locations Suggestion IntelliJ Plugin:
This plugin uses the model from PART 1. It uses inspection to find out comment worthy locations and show them as warning.
When hover over the line, it displays the confidence level for the prediction.


![example](https://user-images.githubusercontent.com/40366759/135456209-e5dd4816-31f6-40f8-8bd9-a577e3f29fcf.png)

### Prerequisites
The model files from part 1, which are "model.onnx", "word_embeddings.txt", "word_to_indexes.csv", "word.vectors.txt", 
should be under the config file of the Plugin (comment_locator-main/build/idea-sandbox/config)

### Limitations
- As explained in the paper, the recall value is too low for the model (approx. 15%). It means, it cannot find the most comment worthy locations. However,
precision is higher around 80 %, so if it suggest something, it is more reliable. 

- It is assumed that the each code chunks seperated by empty lines are coherent itself. So, if developer writes code by putting random epty spaces, the model could not make sense out of it.


### Future Works
 This is an beginning of studies on this subject. There are many directions to be worth to follow in terms of model improvement as well as the Plugin implementation. For example:
- The dataset for the training data can be extended to improve results. 
- The idea can be applied to other languages. However there is no existed dataset for different languages. (Some preliminary analysis done to create such JAVA datasets that can be found under analysis_java_files.ipynb. Java repositories from the link with more than 50 stars  (https://jetbrains.team/p/ccrm/repositories/fl-dataset/files/docs/README.md#download) are investigated.)
- Further step could be also combine this project with projects related to WHAT TO COMMENT, not only WHERE TO COMMENT.

### Relevant Paper& Works 
- https://homes.cs.washington.edu/~mernst/pubs/predict-comments-icse2020-abstract.html

- https://dl.acm.org/doi/10.1145/3434280

- https://github.com/SpringHerald/DeepCommenter


CHANGELOG.md
<!-- Keep a Changelog guide -> https://keepachangelog.com -->

# IntelliJ Platform Plugin Template Changelog

## [Unreleased]
## [0.10.1]
## [0.10.1]
### Added
- Introduced `next` branch in the root repository to make `main` always a stable one

### Changed
- Dependencies (GitHub Actions) - upgrade `actions/cache` to `v2.1.6`
- Trigger GitHub Actions `Build` workflows only on pushes to `main` branch or pull request to avoid duplicated checks

## [0.10.0]
### Changed
- Remove reference to the `jcenter()` from Gradle configuration file
- Update `pluginVerifierIdeVersions` to `2020.2.4, 2020.3.4, 2021.1.2`
- Update `pluginUntilBuild` to include `211.*` (2021.1.*)  
- Dependencies - upgrade `org.jetbrains.kotlin.jvm` to `1.5.10`
- Dependencies - upgrade `detekt-formatting from` to `1.17.1`
- Dependencies - upgrade `io.gitlab.arturbosch.detekt` to `1.17.1`
- Dependencies (GitHub Actions) - upgrade `actions/cache` to `v2.1.5`
- Dependencies (GitHub Actions) - upgrade `actions/checkout` to `v2.3.4`
- Dependencies (GitHub Actions) - upgrade `actions/upload-release-asset` to `v1.0.2`
- Dependencies (GitHub Actions) - upgrade `actions/create-release` to `v1.1.4`
- Upgrade Gradle Wrapper to `7.0.2`

## [0.9.0]
### Added
- `properties` shorthand function for accessing `gradle.properties` in a cleaner way
- Dependabot check for GitHub Actions used in [workflow files](.github/workflows)

### Changed
- Dependencies - upgrade `detekt-formatting from` to `1.16.0` 
- Dependencies - upgrade `io.gitlab.arturbosch.detekt` to `1.16.0` 
- Dependencies - upgrade `org.jetbrains.kotlin.jvm` to `1.4.32` 
- Dependencies (GitHub Actions) - upgrade `actions/upload-artifact` to `v2.2.2`
- Dependencies (GitHub Actions) - upgrade `actions/cache` to `v2.1.4`

### Fixed
- Fix `README.md` file resolution in the `build.gradle.kts`

## [0.8.3]
### Changed
- Dependencies - upgrade `org.jetbrains.intellij` to `0.7.2`
- Dependencies - upgrade `org.jlleitschuh.gradle.ktlint` to `10.0.0`
- Update `platformVersion` to `2020.2.4` for compatibility with macOS Big Sur
- Upgrade Gradle Wrapper to `6.8.3`

## [0.8.2]
### Changed
- Use `-bin` distribution of the Gradle Wrapper
- Upgrade Gradle Wrapper to `6.8.2`
- Update `pluginVerifierIdeVersions` in `gradle.properties` files
- Dependencies - upgrade `org.jetbrains.kotlin.jvm` to `1.4.30`
- Dependencies - upgrade `org.jetbrains.changelog` to `1.1.1`
- Configure the `changelog` Gradle plugin

## [0.8.1]
### Added
- README: Dependencies management section

### Changed
- Upgrade Gradle Wrapper to `6.8`
- Dependencies - upgrade `org.jetbrains.changelog` to `1.0.0`

### Fixed
- Template Cleanup: Escape GitHub username to avoid incorrect characters in class package name
- Template Cleanup: Run `ktlintFormat` task to fix imports order
- GitHub Actions: Use the correct property in the "Upload artifact" step

## [0.8.0]
### Added
- Dependabot integration
- Show `idea.log` logs of the run IDE in the Run console
- README: FAQ section

### Changed
- `build.gradle.kts`: simpler syntax for configuring `KotlinCompile`
- Dependencies - upgrade `org.jetbrains.kotlin.jvm` to `1.4.21`
- Dependencies - upgrade `detekt-formatting` to `1.15.0`
- Dependencies - upgrade `io.gitlab.arturbosch.detekt` to `1.15.0`
- README: Clarify the Java usage in the project
- `pluginVerifierIdeVersions` - upgrade to `2020.1.4, 2020.2.3, 2020.3.1`

### Fixed
- Return `Supplier<@Nls String>` instead of `String` in `MyBundle.messagePointer`

## [0.7.1]
### Changed
- Upgrade Gradle Wrapper to `6.7.1`
- Dependencies - upgrade `org.jetbrains.intellij` to `0.6.5`
- Dependencies - upgrade `org.jetbrains.kotlin.jvm` to `1.4.20`
- Update the base platform version to 2020.1
- Change since/until build to `201-203.*`

## [0.7.0]
### Added
- Predefined Run/Debug Configurations
- Project icon for development purposes

### Changed
- Dependencies - upgrade `org.jetbrains.intellij` to `0.6.3`

## [0.6.1]
### Added
- GitHub Actions - use hash based on `pluginVerifierIdeVersions` in `Setup Plugin Verifier IDEs Cache` step

### Changed
- Use [Kotlin extension function](https://plugins.jetbrains.com/docs/intellij/plugin-services.html#retrieving-a-service) to retrieve the `MyProjectService` in the `MyProjectManagerListener`
- Dependencies - upgrade `org.jetbrains.intellij` to `0.6.2`
- Update `pluginVerifierIdeVersions` in the `gradle.properties` files

## [0.6.0]
### Added
- Integration with [IntelliJ Plugin Verifier](https://github.com/JetBrains/intellij-plugin-verifier) through the [Gradle IntelliJ Plugin](https://github.com/JetBrains/gradle-intellij-plugin#plugin-verifier-dsl) `runPluginVerifier` task
- Cache downloaded IDEs used by Plugin Verifier for the verification

### Changed
- Switch Gradle Wrapper to `-all` to improve the IntelliSense
- Update detekt config to be in line with IJ settings
- Dependencies - upgrade `io.gitlab.arturbosch.detekt` to `1.14.2`
- Dependencies - upgrade `org.jetbrains.intellij` to `0.6.1`
- GitHub Actions - `gradleValidation` update to `gradle/wrapper-validation-action@v1.0.3`
- GitHub Actions - `releaseDraft` update to `actions/download-artifact@v2`

### Removed
- Remove Third-party IntelliJ Plugin Verifier GitHub Action

## [0.5.1]
### Added
- Missing properties in the `gradle.properties` template file

### Changed
- Upgrade Gradle Wrapper to `6.7`
- Dependencies - upgrade `org.jetbrains.changelog` to `0.6.2`

## [0.5.0]
### Added
- Introduced `platformPlugins` property in `gradle.properties` for configuring dependencies to bundled/external plugins

### Changed
- Disable "Release Draft" job for pull requests in the "Build" GitHub Actions Workflow
- Dependencies - upgrade `org.jetbrains.intellij` to `0.5.0`
- Dependencies - upgrade `org.jetbrains.changelog` to `0.6.1`
- Dependencies - upgrade `io.gitlab.arturbosch.detekt` to `1.14.1`
- Dependencies - upgrade `org.jlleitschuh.gradle.ktlint` to `9.4.1`
- Remove LICENSE file during the Template Cleanup workflow

## [0.4.0]
### Added

train_orig_ptl.py
import argparse

from einops import repeat
import pytorch_lightning as pl
from pytorch_lightning.callbacks import ModelCheckpoint
from pytorch_lightning.loggers import TensorBoardLogger

from model import HierarchicalLSTM
from data_processing import (
    read_comment_location_file,
    create_vocabularies,
    read_w2v_models,
    DataModule,
)

parser = argparse.ArgumentParser(description="Train a PyTorch Lightning model")

parser.add_argument(
    "--data_train_path", default="data/split30_word_data/train/data.txt", help="",
)
parser.add_argument(
    "--data_test_path", default="data/split30_word_data/test/data.txt", help="",
)
parser.add_argument(
    "--data_validation_path", default="data/split30_word_data/valid/data.txt", help="",
)

parser.add_argument(
    "--check_point_dir",
    default="model_save/",
    help="Directory for saving model during training",
)

parser.add_argument("--embed_size", default=300, type=int, help="")
parser.add_argument("--max_blocks", default=10, type=int, help="")
parser.add_argument("--max_length_block", default=10, type=int, help="")
parser.add_argument("--max_length_sentence", default=10, type=int, help="")
parser.add_argument("--vocab_size", default=5000, type=int, help="")
parser.add_argument("--embedding_type", default="avg_wembed", help="")
parser.add_argument(
    "--code_embeddings",
    default="data/split30_word_data/code-big-vectors-negative300.bin",
    help="",
)

# Model related arguments #
parser.add_argument("--use_gpu", default=0, type=int, help="1 for GPU usage")
parser.add_argument("--lstm_hidden_size", default=800, type=int, help="")
parser.add_argument("--lstm_num_layers", default=1, help="")
parser.add_argument(
    "--batch_size", default=64, type=int, help="Batch size for training"
)
parser.add_argument("--learning_rate", default=0.001, type=float)
parser.add_argument("--max_epochs", default=1, type=int)
parser.add_argument("--train_batch_size", default=64, type=int)
parser.add_argument("--val_batch_size", default=64, type=int)
parser.add_argument("--test_batch_size", default=1, type=int)
parser.add_argument("--mode", default="train", help="train or predict")
parser.add_argument(
    "--model_for_prediction",
    default="model_save/model-epoch=00-val_loss=0.50.ckpt",
    help="Path for model to predict",
)
parser.add_argument("--word_vectors_file", default="word_vectors.txt")
parser.add_argument("--word_to_indexes", default="all_word_w2i.txt")

args = parser.parse_args()

file_seqs, all_words = read_comment_location_file(
    args.data_train_path, args.max_length_sentence, True,
)
file_seqs_val, _ = read_comment_location_file(
    args.data_validation_path, args.max_length_sentence, True,
)

file_seqs_test, _ = read_comment_location_file(
    args.data_test_path, args.max_length_sentence, True,
)

(wordv, all_wordv, all_word_w2i, all_word_i2w) = create_vocabularies(
    all_words, args.vocab_size
)

input_file = open(args.word_vectors_file, "w")
for element in wordv:
    input_file.write(element + "\n")
input_file.close()

input_file = open(args.word_to_indexes, "w")
for element in all_word_w2i:
    input_file.write(element + " " + str(all_word_w2i[element]) + "\n")
input_file.close()

w2v_models, w2v_dims = read_w2v_models(args.code_embeddings)

data_module = DataModule(
    args,
    w2v_models,
    w2v_dims,
    wordv,
    all_wordv,
    all_word_w2i,
    all_word_i2w,
    file_seqs,
    file_seqs_val,
    file_seqs_test,
)

args.vectors_store = data_module.pretrainedvecs

if args.mode == "train":
    classifier = HierarchicalLSTM(args)

    checkpoint_callback = ModelCheckpoint(
        monitor="val_loss",
        dirpath=args.check_point_dir,
        every_n_val_epochs=1,
        filename="model-{epoch:02d}-{val_loss:.2f}",
        save_top_k=1,
        mode="min",
        save_weights_only=False,
    )

    trainer = pl.Trainer(
        max_epochs=args.max_epochs,
        gpus=args.use_gpu,
        logger=TensorBoardLogger("logs"),
        gradient_clip_val=50,
        gradient_clip_algorithm="norm",
        track_grad_norm=2,
        callbacks=[checkpoint_callback],
    )
    trainer.fit(
        classifier, data_module.train_dataloader(), data_module.val_dataloader()
    )

elif args.mode == "predict":
    if args.model_for_prediction == "none":
        print("Model is not loaded.")
        exit()

    classifier = HierarchicalLSTM(args)
    model = classifier.load_from_checkpoint(
        args.model_for_prediction, kwargs=dict(args=args)
    )
    print(model)
    trainer = pl.Trainer(gpus=args.use_gpu)
    test_loader = data_module.test_dataloader()
    result = trainer.test(model=model, test_dataloaders=test_loader)
    print(result)
